<a href="https://colab.research.google.com/github/ArmandoSep/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

In [2]:
### YOUR CODE STARTS HERE

!wget --no-check-certificate 'https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data'

--2020-04-09 21:55:57--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
  Issued certificate has expired.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-04-09 21:55:58 (132 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [0]:
import pandas as pd
import numpy as np

In [5]:
df = pd.read_csv('house-votes-84.data', names=['party','handicapped-infants','water-project',
                                                 'budget','physician-fee-freeze','el-salvador-aid',
                                                 'religious-groups','anti-satellite-ban',
                                                 'aid-t-contras','mx-aissile','immigration',
                                                 'synfuels','education','right-to-sue','crime','duty-free',
                                                 'south-africa'])

print (df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [6]:
# Replace y to 1, n to 0, ? to NaN

df = df.replace({'y':1,'n':0,'?': np.NaN})
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
# Create a dataset for republicans and democrats

rep = df[df['party']=='republican']
dem = df[df['party']=='democrat']

In [0]:
rep.head()

In [0]:
dem.head()

In [0]:
from scipy import stats

## Bullet 2

Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01

#### 1. Test
**Null Hypothesis**: More than 70% of republicans support this bill. 

$H_0: \mu == \bar(x)$   (\mu is .7)

**Alternative Hypothesis**: 

$H_0: \mu \neq \bar(x)$ 

Republican support is not 70% (it's something other than 70%)

**Confidence Value:** 99%


In [66]:
# Run t test republican
stats.ttest_1samp(rep['south-africa'], .7, nan_policy='omit')

Ttest_1sampResult(statistic=-1.0775937700884006, pvalue=0.2830045640917732)

In [62]:
rep['south-africa'].mean()

0.6575342465753424

#### 2. Test 
**Null Hypothesis**: More than 70% of democrats do not support this bill. 

$H_0: \mu == \bar(x)$   (\mu is .7)

**Alternative Hypothesis**: 

$H_0: \mu \neq \bar(x)$ 

Democrat support is not 70% (it's something other than 70%)

**Confidence Value:** 99%

In [65]:
# Run t test Democrats
stats.ttest_1samp(dem['south-africa'], .7, nan_policy='omit')

Ttest_1sampResult(statistic=12.950439098052623, pvalue=1.0990862796127252e-27)

In [64]:
dem['south-africa'].mean()

0.9351351351351351

Conclusion test 1: Based on the P-Value of 0.283, I reject the null hypothesis. Less than 70% of republicans support this bill (negative statistic).

Conclusion test 2: Based on the P-Value of ~0%, I fail to reject the null hypothesis. More than 70% of democrats support this bill.

### Conclusion
**More democrats support the education bill than republicans.**

## Bullet 3

Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01

#### 1. Test
**Null Hypothesis**: More than 20% of republicans support this bill. 

$H_0: \mu == \bar(x)$   (\mu is .2)

**Alternative Hypothesis**: 

$H_0: \mu \neq \bar(x)$ 

Republican support is not 20% (it's something other than 20%)

**Confidence Value:** 99%

In [53]:
# Run t test republican
stats.ttest_1samp(rep['education'], .2, nan_policy='omit')

Ttest_1sampResult(statistic=24.83772518956985, pvalue=9.882639150170844e-56)

In [36]:
rep['education'].mean()

0.8709677419354839

#### 2. Test 
**Null Hypothesis**: More than 20% of democrats do not support this bill. 

$H_0: \mu == \bar(x)$   (\mu is .2)

**Alternative Hypothesis**: 

$H_0: \mu \neq \bar(x)$ 

Democrat support is not 20% (it's something other than 20%)

**Confidence Value:** 99%

In [52]:
# Run t test Democrats
stats.ttest_1samp(dem['education'], .2, nan_policy='omit')

Ttest_1sampResult(statistic=-2.4817833963283507, pvalue=0.013735856027787562)

In [37]:
dem['education'].mean()

0.14457831325301204

Conclusion test 1: Based on the P-Value of ~0, I fail to reject the null hypothesis. More than 20% of republicans support this bill.

Conclusion test 2: Based on the P-Value of 0.0137, I  reject the null hypothesis. Less than 20% of democrats support this bill (negative statistic).

### Conclusion
**More republicans support the education bill than democrats.**



### Bullet 4

Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)


Null hypothesis: the two samples are different

Alternative: The two samples are similar

(I do not get to choose the null hypothesis. It's always that they're the same in a 2-sample t-test.

Confidence level: 90%

In [71]:
stats.ttest_ind(rep['water-project'],dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

In [72]:
rep['water-project'].mean()

0.5067567567567568

In [73]:
dem['water-project'].mean()

0.502092050209205

**Conclusion**: Based on a P-Value of 0.92, I reject the null hypotesis. The two samples are similar. 

## Stretch Goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then check your Answers using Scipy!